### ISSUE TITLE
[FEATURE] Allow to fit / calibrate conformal models

### ISSUE THREAD

**Issue by swertz** [unit8co/darts#2886]
**Describe the issue linked to the documentation**

It's great that darts provides the possibility to calibrate the forecasts of a model using conformal prediction!

In practice, we are having difficulties using it because the calibration happens every time `predict()` is called (as explained in the [example](https://unit8co.github.io/darts/examples/23-Conformal-Prediction-examples.html)). This requires running an historical forecast on a relatively long timeseries. In a live production setting, where we run many forecasts at regular intervals (15 minutes), this forces us to fetch much more data than we would like, and it makes the forecasting very slow.

It seems to me that the calibration could be done _once_ after the training (using a calibration set that is independent of the training data), so that the _calibrated_ model is stored and used for inference, without having to recalibrate every time a prediction is made.

Of course, this requires the model to be regularly recalibrated, but the model anyway needs to be retrained regularly to account for drifts.

**Additional context**

[crepes](https://github.com/henrikbostrom/crepes) provides an explicit `calibrate()` method that should be called after `fit()` and before `predict()`.

[conformal-tights](https://github.com/superlinear-ai/conformal-tights) is not so clear in the README about what is used for the calibration, but looking at the code, the predictor is calibrated within the `fit()` method and not `predict()`.

[nixtla](https://nixtlaverse.nixtla.io/neuralforecast/docs/tutorials/conformal_prediction.html) also doesn't make it very clear in the documentation, but AFAIU the calibration is done in the training set.

**Additional question**

I might open a separate issue for this, but it seems like `historical_forecast` and `backtest` do not yet support testing conformal calibrated models. Are there any plans to support that?

**Comment by dennisbader** [comment]
Hi @swertz and thanks for raising this feature request. I agree 100% and already thought about adding this :)


My current thoughts:

- We should allow (or maybe then rather enforce? not 100% sure about that yet) to fit the conformal model with `ConformalModel.fit(series, ...)` on the calibration set. This would store a set of parameters that will be applied at prediction to conformalize the intervals.
  - the underlying base-forecasting model will not be fit
- then users can use the `retrain` parameter in historical forecasts to define whether to always fit a new set of parameters (as it is currently the case) or to use the pre-fit parameters.
- for the multi-series case, we could / should add a new model creation parameter: `global_fit` (similar to global fit from our data transformers): Optionally, whether all `TimeSeries` passed to the `fit()` method should be used to fit a *single* set of parameters, or if a different set of parameters should be independently fitted to each provided `TimeSeries`.
  - at the moment `global_fit=True` with `historical_forecasts(retrain=True)` should not be allowed, since the alignment of the series could cause lookahead bias. We're currently working on adding historical forecasts with global retraining (on the intersection of all provided series). After this is implemented, we can allow this case.

How does that sound?

About your additional question:
Not sure I fully understand what you mean :) The examples [here](https://unit8co.github.io/darts/examples/23-Conformal-Prediction-examples.html#Example-3:-Multi-horizon-Forecasts-on-a-Scheduled-Basis-with-valid-Coverage) show how you can perform historical forecasts, backtest, residuals with the conformal models (identical user flow as for regular models).

**Comment by swertz** [comment]
Thanks for the quick reply @dennisbader !

> We should allow (or maybe then rather enforce? not 100% sure about that yet) to fit the conformal model with ConformalModel.fit(series, ...) on the calibration set. This would store a set of parameters that will be applied at prediction to conformalize the intervals.
>
>    the underlying base-forecasting model will not be fit

This sounds good to me. It's nice to have the calibration step explicit.

> then users can use the retrain parameter in historical forecasts to define whether to always fit a new set of parameters (as it is currently the case) or to use the pre-fit parameters.

I don't quite see how this would work in an historical forecast with periodic retraining, since in that case both the "base" model and the calibration model should be re-fit. It seems necessary  to specify not only the `train_length` but also the `calibration_length`, no?

> for the multi-series case, we could / should add a new model creation parameter:

This is not a feature we've been using, I have no opinion on this.

> The examples [here](https://unit8co.github.io/darts/examples/23-Conformal-Prediction-examples.html#Example-3:-Multi-horizon-Forecasts-on-a-Scheduled-Basis-with-valid-Coverage) show how you can perform historical forecasts, backtest, residuals with the conformal models (identical user flow as for regular models).

Thanks for the pointer, we'll check again the exact problem we encountered and open a separate issue if needed!

**Comment by dennisbader** [comment]
Thanks for the answer @swertz, much appreciated :) 

> > We should allow (or maybe then rather enforce? not 100% sure about that yet) to fit the conformal model with ConformalModel.fit(series, ...) on the calibration set. This would store a set of parameters that will be applied at prediction to conformalize the intervals.
> > 
> > the underlying base-forecasting model will not be fit
> 
> This sounds good to me. It's nice to have the calibration step explicit.

From a user perspective, would you have a preference for "enforcing" the fit before predict or leave it optional? Maybe enforcing would be a more clear user flow 🤔  I'm happy about any thought :)

> I don't quite see how this would work in an historical forecast with periodic retraining, since in that case both the "base" model and the calibration model should be re-fit. It seems necessary to specify not only the train_length but also the calibration_length, no?

I meant re-fitting the parameters of the ConformalModel, not the underlying base forecasting model. Fitting the base forecaster would require extra work and complexity which I'm not yet super keen on adding.

**Comment by swertz** [comment]
> From a user perspective, would you have a preference for "enforcing" the fit before predict or leave it optional? Maybe enforcing would be a more clear user flow 🤔 I'm happy about any thought :)

Would the un-fitted conformal wrapper act like a no-op and return the base model's raw forecasts? If yes, why not, as long as the docs clearly state that the wrapper needs to be fitted separately for calibration to work :) 

(Of course one could also simply not add the conformal wrapper, if the goal is to compare to uncalibrated forecasts)

If the behaviour of the un-fitted ConformalModel is undefined, then I would absolutely vote for enforcing its fit! 

> I meant re-fitting the parameters of the ConformalModel, not the underlying base forecasting model. Fitting the base forecaster would require extra work and complexity which I'm not yet super keen on adding.

Ah, I see! What I had in mind was that the historical forecast/backtest procedure with retraining would both retrain the base model, and recalibrate the conformal wrapper. Otherwise it becomes difficult to evaluate the model's accuracy in a way that's consistent with the production environment (periodic retraining and recalibration).

This actually was my additional point in the ticket: currently, we have to write an "explicit" backtest loop, to retrain and recalibrate the model at regular intervals on the historical data.

**Comment by dennisbader** [comment]
> Would the un-fitted conformal wrapper act like a no-op and return the base model's raw forecasts? If yes, why not, as long as the docs clearly state that the wrapper needs to be fitted separately for calibration to work :)

Let's say here we would be talking about **not enforcing fit before predict**. Here I meant that the ConformalModel would behave as before. So you could call predict() directly and it would calibrate on the past of the `series` that you pass. Now I realize that this would be a bit ambiguous if fit was actually called before. Would it then use the pre-fit parameters or re-calibrate on the past of the series? So I guess enforcing fit before predict makes it more clear.

> Ah, I see! What I had in mind was that the historical forecast/backtest procedure with retraining would both retrain the base model, and recalibrate the conformal wrapper. Otherwise it becomes difficult to evaluate the model's accuracy in a way that's consistent with the production environment (periodic retraining and recalibration).

I see your point. Adding support for this would require a lot of things to be added / added complexity though:

- there will be an ambiguity for the `retrain` parameter whether it means retrain the base forecaster or the re-calibrate the conformal model. Maybe we would have to introduce a new param `recal` to control that? At the moment `recal=True` for conformal models as the recalibrate every step. `recal=False` would mean it uses the parameters fitted when calling `ConformalModel.fit`.
- What about other parameters like `start`, would it be the start of the base forecaster or the conformal model (this is coupled with `retrain` unfortunately)? Before this didn't matter.
- we should introduce `val_length` to ForecastingModel.historical_forecasts (FHF) to allow early stopping for global models (this is already being worked on). During FHF, this would automatically fetch `val_length` points after the end of the respective training set and pass it as `val_series` to `model.fit()`. In that case we should not produce historical forecasts during the window of the validation set.
- should we introduce `cal_length` to ConformalModel.historical_forecasts (CHF) and remove it from ConformalModel.__init__? This would be more consistent with how we do it for the regular/base forecasting models, where in FHF you can define `train_length` and during ForecastingModel.fit you can define `max_samples_per_ts` to control how many samples are used to train the model.
- there will probably be other points that I couldn't think of right now.

A positive point about allowing this would be that we could in theory also allow any base forecaster (not only global models) since retrain would be allowed. But that again introduces complexity during `ConformalModel.fit()` since the user should get control over how the calibration set is generated using the base forecaster's historical forecast 😆

**Comment by swertz** [comment]
> So I guess enforcing fit before predict makes it more clear.

Yes, I think that's better than ambiguity!

For the backtest, I understand the complexity... I see essentially six scenarios, but I also don't have the full picture:

1. Base model not retrained
  a. No recalibration (use calibrated model as is)
  b. "Live" recalibration, at every prediction stride. Requires specifying the `cal_length`; the calibration sets are located right before the test sets.
  c. Periodic recalibration (same behaviour as the current `retrain` argument). Requires specifying the `cal_length` and `cal_frequency`. The calibration sets are located after the training sets and before the test sets (introducing a gap).
2. Base model retrained
  a. "Live" recalibration, at every prediction stride. Requires `cal_length`.
  b. Periodic recalibration with same frequency as the base model retraining. Requires `cal_length`.
  c. Periodic recalibration with higher frequency than the base model retraining. Requires `cal_length` and `cal_frequency`.

Perhaps the "live" recalibration can be handled as a special case of the `cal_frequency`...

I also agree with your point that it's more consistent to have `cal_length` in `ConformalModel.fit`.

Perhaps this complexity points to the need of having a toolbox for the backtesting/historical forecast that is separate from the models themselves (i.e. not anymore model methods)? One could then have different backtesting strategies, and those strategies accept any model as input. There could then be a specific strategy for calibrated models, that handles the retraining and recalibration without adding complexity to "regular" backtesting.
